frappe/model/document.py

```diff
@@ -88,35 +88,28 @@ def __init__(self, *args, **kwargs):
 		If DocType name and document name are passed, the object will load
 		all values (including child documents) from the database.
 		"""
-		self.doctype = self.name = None
-		self._default_new_docs = {}
+		self.doctype = None
+		self.name = None
 		self.flags = frappe._dict()
 
-		if args and args[0] and isinstance(args[0], str):
-			# first arugment is doctype
-			if len(args)==1:
-				# single
-				self.doctype = self.name = args[0]
-			else:
+		if args and args[0]:
+			if isinstance(args[0], str):
+				# first arugment is doctype
 				self.doctype = args[0]
-				if isinstance(args[1], dict):
-					# filter
-					self.name = frappe.db.get_value(args[0], args[1], "name")
-					if self.name is None:
-						frappe.throw(_("{0} {1} not found").format(_(args[0]), args[1]),
-							frappe.DoesNotExistError)
+
+				if len(args) == 1:
+					# single
+					self.name = self.doctype
 				else:
 					self.name = args[1]
+					self.flags.for_update = kwargs.get("for_update")
```

Comment:
This should probably be moved a bit higher since `for_update` doesn't seem to work for singles (and never did!). Also, doesn't work for Dict/List filters either.

Reply:
`for_update` _is_ implemented for Dict/List filters. This is because they are a part of this else block. `self.name` here could be `dict` / `list` filters as well, not just your vanilla name.

---

Implemented `for_update` for singles in https://github.com/frappe/frappe/pull/16474/commits/e73c5526328af7302080e7f23c7943bd0918d834.

Tested locally:

![image](https://user-images.githubusercontent.com/16315650/161229095-f4a35690-b22b-4f04-8e18-23bd5d9996f9.png)